src/ex3/VariableInitVisitor.java

```diff
@@ -5,172 +5,218 @@
+		program.mainClass().accept(this);
+		for (ClassDecl classdecl : program.classDecls()) {
+			classdecl.accept(this);
+		}
+	}
+
+	@Override
+	public void visit(ClassDecl classDecl) {
+		for (var methodDecl : classDecl.methoddecls()) {
+			methodDecl.accept(this);
+		}
+	}
+
+	@Override
+	public void visit(MainClass mainClass) {
+//    	mainClass.argsName();
+		mainClass.mainStatement().accept(this);
+	}
+
+	@Override
+	public void visit(MethodDecl methodDecl) {
+//    	for (var formal : methodDecl.formals()) {
+//            formal.accept(this);
+//        }
+		for (var varDecl : methodDecl.vardecls()) {
+			varDecl.accept(this);
+		}
+		for (var stmt : methodDecl.body()) {
+			stmt.accept(this);
+		}
+		methodDecl.ret().accept(this);
+	}
+
+	@Override
+	public void visit(FormalArg formalArg) {
+//    	formalArg.name();
+	}
+
+	@Override
+	public void visit(VarDecl varDecl) {
+		lattice.add(varDecl.name(), initKind.ff);
+	}
+
+	@Override
+	public void visit(BlockStatement blockStatement) {
+		for (var s : blockStatement.statements()) {
+			s.accept(this);
+		}
+	}
+
+	@Override
+	public void visit(IfStatement ifStatement) {
+		Lattice thenLat, elseLat;
+		thenLat = new Lattice();
+		elseLat = new Lattice();
+		thenLat.copy(lattice);
+		elseLat.copy(lattice);
+		ifStatement.cond().accept(this);
+		lattice = thenLat;
+		lattice.assign("y");
+		ifStatement.thencase().accept(this);
+		lattice = elseLat;
+		ifStatement.elsecase().accept(this);
+		lattice.join(thenLat);
+	}
+
+	@Override
+	public void visit(WhileStatement whileStatement) {
+		Lattice origLat = new Lattice();
+		origLat.copy(lattice);
+		whileStatement.cond().accept(this);
+		whileStatement.body().accept(this);
+		lattice.join(origLat);
+	}
+
+	@Override
+	public void visit(SysoutStatement sysoutStatement) {
+		sysoutStatement.arg().accept(this);
+	}
+
+	@Override
+	public void visit(AssignStatement assignStatement) {
+		assignStatement.rv().accept(this);
+
+		lattice.assign(assignStatement.lv());
+	}
+
+	@Override
+	public void visit(AssignArrayStatement assignArrayStatement) {
+		assignArrayStatement.rv().accept(this);
+		assignArrayStatement.index().accept(this);
+
+		lattice.assign(assignArrayStatement.lv());
+	}
+
+	@Override
+	public void visit(AndExpr e) {
+		e.e1().accept(this);
+		e.e2().accept(this);
+	}
+
+	@Override
+	public void visit(LtExpr e) {
+		e.e1().accept(this);
+		e.e2().accept(this);
+	}
+
+	@Override
+	public void visit(AddExpr e) {
+		e.e1().accept(this);
+		e.e2().accept(this);
+	}
+
+	@Override
+	public void visit(SubtractExpr e) {
+		e.e1().accept(this);
+		e.e2().accept(this);
+	}
+
+	@Override
+	public void visit(MultExpr e) {
+		e.e1().accept(this);
+		e.e2().accept(this);
+	}
+
+	@Override
+	public void visit(ArrayAccessExpr e) {
+		e.indexExpr().accept(this);
+	}
+
+	@Override
+	public void visit(ArrayLengthExpr e) {
 
-    private final ClassMapping classMap;
+	}
+
+	@Override
+	public void visit(MethodCallExpr e) {
 
```

Comment:
Should probably check that all args have been initialized

Reply:
yes I guess, I added that check